Fix parameter parsing for 'array of' types

fourls opened the pull request:
Currently, the parameter parsing introduced in #161 does not handle instances where a parameter type is multiple tokens, e.g. `array of String`, `file of TObject`, `set of Integer`. This can lead to mangled IDs in HTML output, where `procedure A(MyArr: array of String)` has the same ID as `procedure A(MyStr: string)`.

This PR fixes this issue by correctly parsing multi-token parameter types, and removing any spaces in the parameter types when generating the signature for a routine.

I've also added a test case and regenerated output.

michaliskambi commented on source/component/PasDoc_Parser.pas:
```diff
@@ -898,20 +898,36 @@ procedure TParser.ParseRoutine(out M: TPasRoutine;
+  begin
+    if Token.MyType = TOK_IDENTIFIER then
+      Result := True
+    else if Token.MyType = TOK_KEYWORD then begin
```
Put `begin` on the same indentation as `end`, and `if` at the beginning, so

```
    else 
    if Token.MyType = TOK_KEYWORD then 
    begin
       ...
    end else
```

I know, we have various styles in PasDoc, but going forward I'd like to make it consistent -- and I'd say to follow CGE coding style, https://castle-engine.io/coding_conventions .

michaliskambi commented on source/component/PasDoc_Parser.pas:
```diff
@@ -898,20 +898,36 @@ procedure TParser.ParseRoutine(out M: TPasRoutine;
+      Result := True
+    else if Token.MyType = TOK_KEYWORD then begin
+      Text := AnsiLowerCase(Token.Data);
+      Result := (Text = 'string') or (Text = 'array') or (Text = 'set') or (Text = 'of') or (Text = 'file');
```
No need to compare text here, instead use ready `Token.KeyWord`. So this is better:

```pascal
Result := Token.KeyWord in [KEY_STRING, KEY_ARRAY, KEY_SET, ...];
```

( I see previous code also used `AnsiLowerCase(T.Data) = 'string')`,I didn't catch it earlier, sorry. )

fourls replied:
I didn't realise that functionality existed, changed :)

michaliskambi commented on source/component/PasDoc_Items.pas:
```diff
@@ -2859,11 +2859,24 @@ procedure TPasRoutine.Serialize(const ADestination: TStream);
   else
     Result := Name;
+
+  I := 1;
```
There is already a very similar code in `NormalizeSignature` in `TPasRoutines.FindListItem`, that removes spaces from string.

Let's make a routine in `PasDoc_Utils` that can be used for this, like `function SRemoveChars(const S: String; const C: Char): String;`

michaliskambi reviewed (changes requested):
Thank you! Overall looks cool, I found some minor things to improve :)

fourls commented:
Thanks for reviewing Michalis! I've fixed up those issues.

I also realised that the `out` keyword was not registered as a keyword, resulting in some routine signatures being incorrect as it assumed the identifier was a parameter name. For example, `procedure Foo(const MyStr: string)` had the generated HTML id `#Foo-string-string-` instead of the correct `#Foo-string-`. I've fixed that up too.

michaliskambi commented:
Hm, I just realized the automatic tests fail for FPC 3.2.0. For some reason, when PasDoc is compiled with FPC 3.2.0, it cannot pass the `ok_array_parameters.pas` testcase, reporting:

```
Warning[1]: Could not resolve link "Foo(array of string, array of Integer, file of TObject)" (from description of "ok_array_parameters.Bar")
```

Works OK with FPC 3.2.2 and very latest FPC 3.3.1 from GitLab as of today. Weird.

GitHub Actions reported this failure, but I didn't pay attention :), blaming it on GitHub outage problems today ( https://twitter.com/githubstatus ).

@fourls Can you debug why it happens, i.e. why FPC 3.2.0 doesn't produce a valid code? I looked at the code, but don't see anything that would be invalid for FPC 3.2.0.

fourls commented:
No worries, I've had a look and figured out the issue - see #182 :)
